proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+The GHC branch that implements the changes described in the proposal is available [here](https://gitlab.haskell.org/tweag/ghc/-/tree/tbagrel1/dps-compact-regions-prims?ref_type=heads) ([diff](https://gitlab.haskell.org/tweag/ghc/-/compare/master...tbagrel1%2Fdps-compact-regions-prims?from_project_id=1)).
+
+## Proposed Change Specification
+
+1. [Allocate hollow data constructor in region: RTS primop `compactAddHollow#`](#allocate-hollow-data-constructor-in-region-rts-primop-compactaddhollow)
+2. [Retreiving the info table pointer: primtype `InfoTablePtrOf#` and compile-time primop `reifyInfoTablePtr#`](#retrieving-the-info-table-pointer-primtype-infotableptrof-and-compile-time-primop-reifyinfotableptr)
+3. [Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family](#obtain-data-constructor-name-as-a-symbol-liftedctortosymbol-built-in-type-family)
+
+### Allocate hollow data constructor in region: RTS primop `compactAddHollow#`
+
+```haskell
+compactAddHollow#
+  :: Compact#         -- ^ the compact region where the constructor heap object will be added
+  -> Addr#            -- ^ the info table pointer that will be set as header for the new constructor heap object
+  -> State# RealWorld -- ^ state token
+  -> (# State# RealWorld
+      , a             -- ^ the allocated constructor heap object
+      , Word#         -- ^ size of the header of the allocated heap object, in bytes
+      , Word#         -- ^ size of the words, in bytes, on the platform
+     #)
+```
+
+This primitive must be implemented in the RTS (in `rts/Compact.cmm`). The behavior of its Cmm code should correspond to these macro steps:
+
+```c
+stg_compactAddHollowzh(
+  P_ compact, // the compact region where the constructor heap object will be added
+  W_ info     // the info table pointer that will be set as header for the new constructor heap object
+) {
+  [...]
+  // find address where to store the pointer to the new constructor object
+  pp = compact + SIZEOF_StgHeader + OFFSET_StgCompactNFData_result;
+  // reuse existing macro to allocate inside the compact region
+  ALLOCATE(compact, size, p, new_ctor_obj, tag);
+  // write the supplied info table pointer
+  SET_HDR(new_ctor_obj, info, CCS_SYSTEM);
+  // store the pointer to the new constructor in the return location
+  P_[pp] = new_ctor_obj;
+  [...]
+
+  return (P_[pp], SIZEOF_StgHeader, SIZEOF_W);
+}
+```
+
+`compactAddHollow#` allocates space for a new constructor heap object in the compact region, but doesn't set any of its fields (so it is _hollow_). It only sets its header with the supplied info table pointer. The user is responsible for setting the fields of the constructor heap object before reading it. The primitive returns the newly allocated constructor heap object, but also the size of the header of the object, and the size of the words on the platform so that the user can later set the fields of the returned object.
```

Comment:
```suggestion
A new primitive, `compactAddHollow#`, which allocates space for a new constructor heap object in the compact region, but doesn't set any of its fields (so it is _hollow_). It only sets its header with the supplied info table pointer. The user is responsible for setting the fields of the constructor heap object before reading it. The primitive returns the newly allocated constructor heap object, but also the size of the header of the object, and the size of the words on the platform so that the user can later set the fields of the returned object.
```

Move to the top of the section. Exercise: remove all the occurrences of the word ”it” in this paragraph.

By the way, did you look at successful proposals which add primops, such as: [Delimited continuation primops](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0313-delimited-continuation-primops.rst) or [sumToTag#](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0369-sum-to-tag.rst)? For inspiration about the style and tone.